File: python/lesson 3/longest_substring.py
# https://leetcode.com/problems/longest-substring-without-repeating-characters

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        left = right = result = 0
        map = dict()

        while right < len(s):
            if s[right] in map:
                left = max(left, map[s[right]] + 1)

            map[s[right]] = right
            result = max(result, right - left + 1)
            right += 1

        return result

File: kotlin/lesson 3/LongestSubstring.kt
// https://leetcode.com/problems/longest-substring-without-repeating-characters/

class Solution {
    fun lengthOfLongestSubstring(s: String): Int {
        var left = 0
        var right = 0
        var result = 0
        val map = mutableMapOf<Char, Int>()

        while (right < s.length) {
            if (map.containsKey(s[right])) {
                left = Math.max(left, map[s[right]]!! + 1)
            }

            map[s[right]] = right
            result = Math.max(result, right - left + 1)
            right++
        }

        return result
    }
}
